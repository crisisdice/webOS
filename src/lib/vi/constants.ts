export enum VI_MODE {
  VISUAL,
  INSERT,
  COMMAND,
  VISUAL_BLOCK,
  VISUAL_LINE,
}
