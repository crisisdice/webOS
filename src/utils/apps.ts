export enum APPS {
  SH,
  VI,
}
