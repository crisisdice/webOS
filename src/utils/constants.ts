export const EMPTY = ''

export const ENV = 'ENV'

export const FS = 'FS'

export const PWD = 'PWD'

export const HOME = 'HOME'

export const ARROW_KEYS = ['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight']
