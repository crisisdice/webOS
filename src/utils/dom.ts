export const refocus = (e: Event) =>
  window.setTimeout(() => (e.target as HTMLInputElement).focus(), 0)
