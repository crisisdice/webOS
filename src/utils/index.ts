export * from './dom'
export * from './state'
export * from './constants'
export * from './apps'
