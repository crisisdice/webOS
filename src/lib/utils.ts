import type { Line } from './types'
import { DASH } from './constants'

export const separate = (args: string): { name: string; path: string } => {
  if (args.indexOf('/') === -1) return { path: '.', name: args }
  const tokens = args.split('/')
  const path = tokens.slice(0, tokens.length - 1).join('/')
  return { name: tokens.at(-1), path }
}

export const refocus = (e: Event) =>
  window.setTimeout(() => (e.target as HTMLInputElement).focus(), 0)

export const lineToString = ({ PRECARET, CARET, POSTCARET }: Line) => {
  return `${PRECARET}${CARET === DASH ? '' : CARET}${POSTCARET}`
}
