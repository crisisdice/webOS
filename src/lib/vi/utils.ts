import type { Line, ViState } from '../../types'
import { write, stat } from '../fs'
import { separate } from '../path'
import { EMPTY_LINE, INITAL_STATE, bufferToFile, lineToString, EMPTY } from '../../utils'
import { VI_MODE } from './constants'

function parseCmd(input: Line) {
  const commandTokens = lineToString(input).split(' ')
  const cmd = commandTokens?.[0] ?? ''
  const args = commandTokens.slice(1)

  const cmds = cmd.split('')
  const override = cmds.includes('!')

  return { cmds, args, override }
}

function writeFile({ args, STATE }: { args: string[]; STATE: ViState }): ViState {
  let { FILE_NAME } = STATE
  const { BUFFER } = STATE
  FILE_NAME = FILE_NAME ?? args?.[0]

  if (!FILE_NAME) return report({ STATE, MESSAGE: 'No file name' })

  write({ ...separate(FILE_NAME), obj: bufferToFile(BUFFER) })
  return { ...STATE, COMMAND_LINE: EMPTY_LINE, MODE: VI_MODE.VISUAL, FILE_NAME }
}

function checkSaveStatus({ STATE }: { STATE: ViState }): boolean {
  const { FILE_NAME, BUFFER } = STATE
  const { exists, obj } = stat(FILE_NAME)
  const file = bufferToFile(BUFFER)

  return !exists ? file === EMPTY : obj === file
}

function report({ STATE, MESSAGE }: { STATE: ViState; MESSAGE: string }) {
  return { ...STATE, MESSAGE, COMMAND_LINE: EMPTY_LINE }
}

export function vimCommands(STATE: ViState) {
  const { COMMAND_LINE, OLD_COMMANDS } = STATE
  const { cmds, args, override } = parseCmd(COMMAND_LINE)

  for (const cmd of cmds) {
    if (cmd === 'w') {
      STATE = writeFile({ args, STATE })
      continue
    }

    if (cmd === 'q') {
      if (!checkSaveStatus({ STATE }) && !override) {
        return report({ STATE, MESSAGE: 'No write since last change' })
      }
      return { ...INITAL_STATE, OLD_COMMANDS }
    }

    return { ...STATE, COMMAND_LINE: EMPTY_LINE, MODE: VI_MODE.VISUAL }
  }
  return { ...STATE, COMMAND_LINE: EMPTY_LINE, MODE: VI_MODE.VISUAL }
}
