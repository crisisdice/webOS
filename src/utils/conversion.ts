import type { Line, ViState } from '../types'
import { EMPTY, PWD, EMPTY_BUFFER } from './constants'
import { getEnv, stat } from '../lib'

export const lineToString = ({ PRECARET, CARET, POSTCARET }: Line) => {
  return `${PRECARET}${CARET === null ? '' : CARET}${POSTCARET}`
}

export const stringToLine = (line: string, x: number, CARET_WIDTH = 1): Line => {
  const PRECARET = line.slice(0, x)
  const EOL = x === line.length
  const CARET = EOL || CARET_WIDTH === 0 ? null : line[x]
  const POSTCARET = EOL ? EMPTY : line.slice(x + CARET_WIDTH)

  return {
    PRECARET,
    CARET,
    POSTCARET,
    EOL,
    CARET_WIDTH,
  }
}

export function fileToBuffer(fileName: string | null): {
  BUFFER: ViState['BUFFER']
  FILE_NAME: string | null
} {
  const { exists, obj, path, isDirectory } = stat('', fileName)

  if (!exists || isDirectory) {
    return { BUFFER: EMPTY_BUFFER, FILE_NAME: fileName ? `${getEnv(PWD)}/${fileName}` : null }
  }
  if (obj === EMPTY) {
    return { BUFFER: EMPTY_BUFFER, FILE_NAME: path }
  }

  const file = (obj as string).split('\n')

  const BUFFER_PRE = []
  const LINE = stringToLine(file[0], 0)
  const BUFFER_POST = file.slice(1)

  return { BUFFER: { BUFFER_PRE, BUFFER_POST, LINE }, FILE_NAME: path }
}
