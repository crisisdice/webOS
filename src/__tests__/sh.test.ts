import { run, fill } from './helpers'
import { DOWN, ENTER, LEFT, RIGHT, UP } from './constants'
import { init } from '../utils'

const tests = [
  [...fill(3, LEFT), ...fill(3, RIGHT)],
  ['g', LEFT],
  [...'hello'.split(''), ...fill(5, LEFT), ...fill(6, RIGHT)],
  [...'echo'.split(''), ENTER, ...'ls'.split(''), ENTER, UP, UP, DOWN],
]

beforeEach(() => init())

for (const sequence of tests) {
  run(sequence)
}
