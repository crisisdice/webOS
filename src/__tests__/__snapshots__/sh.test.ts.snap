// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LEFT 3><RIGHT 3> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`<LEFT 3><RIGHT 3> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`<LEFT 3><RIGHT 3> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`echo<ENTER>ls<ENTER>ArrowUpArrowUpArrowDown 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "l",
    "CARET_ACTIVE": true,
    "POSTCARET": "s",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": 0,
  "NAME": 0,
  "OLD_COMMANDS": [
    {
      "cmd": "echo",
      "stdout": "",
      "usr": "guest",
      "wd": "/home/guest",
    },
    {
      "cmd": "ls",
      "stdout": ".bashrc .history",
      "usr": "guest",
      "wd": "/home/guest",
    },
  ],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`echo<ENTER>ls<ENTER>ArrowUpArrowUpArrowDown 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`echo<ENTER>ls<ENTER>ArrowUpArrowUpArrowDown 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`g<LEFT 1> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "g",
    "CARET_ACTIVE": true,
    "POSTCARET": "",
    "PRECARET": "",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`g<LEFT 1> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`g<LEFT 1> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;

exports[`hello<LEFT 5><RIGHT 6> 1`] = `
{
  "COMMAND_LINE": {
    "CARET": "_",
    "CARET_ACTIVE": false,
    "POSTCARET": "",
    "PRECARET": "hello",
  },
  "CONTROL_DOWN": false,
  "DOWN_MAPPING": [Function],
  "HISTORY_INDEX": -1,
  "NAME": 0,
  "OLD_COMMANDS": [],
  "PWD": "/home/guest",
  "UP_MAPPING": [Function],
  "USER": "guest",
}
`;

exports[`hello<LEFT 5><RIGHT 6> 2`] = `"{"home":{"guest":{".bashrc":"#!/usr/sh",".history":""},"root":{}}}"`;

exports[`hello<LEFT 5><RIGHT 6> 3`] = `"{"PWD":"/home/guest","HOME":"/home/guest"}"`;
